File: 3-1. Algorithm Analysis and Design/Lab/Lab 3/3-1. 169. Majority Element.java
class Solution {
    public int majorityElement(int[] nums) {
        return helper(nums, 0, nums.length)[0];
    }

    private int[] helper(int[] nums, int l, int r) {
        if (l == r - 1) {
            return new int[]{nums[l], 1};
        }
        int mid = l + (r - l) / 2;
        int[] leftResult = helper(nums, l, mid);
        int[] rightResult = helper(nums, mid, r);

        int maj, extra;
        if (leftResult[0] == rightResult[0]) {
            maj = leftResult[0];
            extra = leftResult[1] + rightResult[1];
        } else if (rightResult[1] > leftResult[1]) {
            maj = rightResult[0];
            extra = rightResult[1] - leftResult[1];
        } else {
            maj = leftResult[0];
            extra = leftResult[1] - rightResult[1];
        }
        return new int[]{maj, extra};
    }
}

File: 3-1. Algorithm Analysis and Design/Lab/Lab 3/3-1. 169. Majority Element.py
import unittest
import os
import json

absolute_path = os.path.abspath(__file__)
file_path = os.path.join(os.path.dirname(absolute_path), "3-1.txt")

class Solution:
    def majorityElement(self, nums) -> int:
        return self.helper(nums, 0, len(nums))[0]

    def helper(self, nums, l, r):
        if l == r - 1:
            return nums[l], 1
        mid = l + (r - l) // 2
        maj_left, extra_left = self.helper(nums, l, mid)
        maj_right, extra_right = self.helper(nums, mid, r)

        if maj_right == maj_left:
            maj = maj_left
            extra = extra_left + extra_right
        elif extra_right > extra_left:
            maj = maj_right
            extra = extra_right - extra_left
        else:
            maj = maj_left
            extra = extra_left - extra_right
        return maj, extra

class TestMajorityElement(unittest.TestCase):

    def test_majority_element(self):
        with open(file_path, 'r') as f:
            line = f.readline().strip()
            arr = json.loads(line)
        solution = Solution()
        self.assertEqual(solution.majorityElement(arr), 3)

if __name__ == "__main__":
    unittest.main(verbosity=2)
